src/bytes.rs

```diff
@@ -193,6 +267,90 @@ pub trait Bytes<A> {
     /// Part of the data may have been copied nevertheless.
     fn read_slice(&self, buf: &mut [u8], addr: A) -> Result<(), Self::E>;
 
+    /// Returns a reference that allows atomic operations for `T` at the specified address.
```

Comment:
I'm hesitating whether the atomic related interfaces should be hosted by a dedicated trait.
Per my understanding, we have three types of memory access modes,
1) read/write in byte stream mode.
2) read/write in naturally aligned mode (word/double word/quad word).
3) read/write/swap/compare_and_exchange in naturally aligned atomic mode with memory ordering.

Bytes::{read/write/read_slice/write_slice/read_from/write_to/read_exact_from/write_all_to} should access memory in byte stream mode.
But Bytes::{read_obj/write_obj} should access memory in naturally aligned mode.
So should we move read_obj/write_obj/atomic_ref/write_atomic/read_atomic out of the Bytes trait?

Reply:
I think both the name and some of the doc comments (i.e. the one about *Candidates which may implement this trait include:...*) are a bit misleading for `Bytes` (in my mind I've always used something along the lines of `RandomAccessMemory`). Your overview of access types sounds reasonable, and what's forcing us right now to explicitly turn operations like `read_obj` into `memcpy`s is cross region access (I'm working on a proposal to address that, and will post something later this week).

To answer your question, I'm not totally sure what the best place for these operations is in the long term; having them in `Bytes` seems like the best option for now. I think we still need a bit more time to figure out how the stable `vm-memory` interface looks like, and we should address some of the existing concerns first.